### ISSUE TITLE
Crash Tag Team Racing Freezes (microVU1: Program cache limit reached)

### ISSUE THREAD

**Issue by Kristijan1001** [PCSX2/pcsx2#2903]
PCSX2 version:

v1.5.0-dev-3015-g936f66d70

Plugins used:
OpenGL Hardware

Description of the issue:
Issue happens in the lobby of Crash Tag Team Racing you get short freezes after every minute, game is playable but after some time fps drop to around 20fps and it freezes after that it unfreezes and it playable again for a bit and so on .

________________________________________

This was attempted to be fixed last year with a game patch which resulted in hard freeze of the game which was reverted, there has been Issue open on this which has been closed, is this something impossible to fix ?

https://github.com/PCSX2/pcsx2/issues/1785

I guess @FlatOutPS2 closed that issue thinking it would be fixed with the game patch but afterwards it created hard freezes instead so it was reverted back.

https://github.com/PCSX2/pcsx2/issues/2366

How to reproduce the issue:
Go to the lobby of the game Crash Tag Team Racing.

Last known version to work:
None

PC specifications:
CPU: Intel i5 3330 2.4GHz
GPU: Nvidia GeForce 1050Ti
OS: Windows 10 64 Bit

**Comment by MrCK1** [comment]
Ok. Let me get this straight. If you disable Gamefixes, you will have *short* pauses but not a hard crash correct?

**Comment by willkuer** [comment]
@prafullpcsx2 is the master of fixing unplayable games. Maybe he has this one and knows how to fix the buggy pnach. Not sure whether he is able to fix such obscure errors.

**Comment by Kristijan1001** [comment]
@MrCK1 yes that is correct, sorry I don’t know how i misstyped that in the op. I think the gamefix was removed a year ago if i am not mistaken cause id made the game completely unplayable.

**Comment by prafullpcsx2** [comment]
> 
> 
> @prafullpcsx2 is the master of fixing unplayable games. Maybe he has this one and knows how to fix the buggy pnach. Not sure whether he is able to fix such obscure errors.

This one doesn't have a pnach. It was using scarface hack (which avoids constant recompilation). That hack was meant only for scarface so doesn't really work correctly with this and results in hard locks. Dev builds won't hard lock but crawl down to unplayable speeds.

Coming back to the game I don't get any short pauses with this. It seems to run pretty good. OP mentions it happens in lobby, so may be it's multiplayer specific? 

What's actually happening in the game? 

Whenever there are coins on screen mvu1 cache keeps increasing. Lots of coins on screen means it happens very quickly. By default mvu1 cache size is 64 mb and it resets after reaching 100%. Issue is when it gets close to 64mb things starts to slow down a lot. On really bad specs pc (like mine) even making a savestate at that time will give out of memory error. Once mvu cache resets to 0 speeds come back to normal.

What can be done?

(a) Use Supervu for VU1. This may result in some errors (needs testing). (Recommended).

or

(b) Use lots of savestates. Keep watching the console and whenever mvu cache reaches dangerously high level make a savestate. As soon as you load that savestate mvu cache will again be down to 0%.

**Comment by Kristijan1001** [comment]
> Coming back to the game I don't get any short pauses with this. It seems to run pretty good. 

@prafullpcsx2 
If you are in a normal lobby , like you explained the game starts slowing down more and more. Then it comes to a point where just before it resets the cache it freezes for a few seconds and then its back to 60FPS. Here is a video of it i recorded it way better to show than to explain 😃 

https://youtu.be/8GSRiJReNx4

**Comment by prafullpcsx2** [comment]
In the console window enable "Dev/Verbose" then you can see the cache filling up.

**Comment by Kristijan1001** [comment]
@prafullpcsx2 The whole emulator crashes after 30 Seconds of gametime. 

`ulti Threaded Fileio module.(99/11/15) 
iop heap service (99/11/03)
loadmodule: fname cdrom0:¥IRX¥NETCNF.IRX;1 args 17 arg icon=a
loadmodule: id 25, ret 2
loadmodule: fname cdrom0:¥IRX¥EZNETCNF.IRX;1 args 0 arg 
loadmodule: id 26, ret 2
(UpdateVSyncRate) Mode Changed to NTSC.
(UpdateVSyncRate) FPS Limit Changed : 59.94 fps
Frame buffer size set to  640x1280 (2560x5120)
loadmodule: fname cdrom0:¥IRX¥SIO2MAN.IRX;1 args 0 arg 
loadmodule: id 27, ret 0
loadmodule: fname cdrom0:¥IRX¥PADMAN.IRX;1 args 0 arg 
loadmodule: id 28, ret 0
loadmodule: fname cdrom0:¥IRX¥MCMAN.IRX;1 args 0 arg 
loadmodule: id 29, ret 2
loadmodule: fname cdrom0:¥IRX¥MCSERV.IRX;1 args 0 arg 
loadmodule: id 30, ret 2
loadmodule: fname cdrom0:¥IRX¥LIBSD.IRX;1 args 0 arg 
loadmodule: id 31, ret 0
loadmodule: fname cdrom0:¥IRX¥SDRDRV.IRX;1 args 0 arg 
SDR driver version 4.0.1 (C) SCEI
 Exit rsd_main 
loadmodule: id 32, ret 2
loadmodule: fname cdrom0:¥IRX¥AUDIO.IRX;1 args 0 arg 
loadmodule: id 33, ret 0

+-----------------------------------------+
| AUDIO: IopLoadServer Init:              |
+------------------------------+----------+
| Version                      |    0x145 |
| Fill Spu With Garbage        |    false |
| Debug Channel Enabled        |    false |
| DVD Transfer Buffer Bytes    |    32768 |
| Memory Transfer Buffer Bytes |    32768 |
| UnInterleave Buffer Bytes    |    32768 |
| Clear Buffer Bytes           |     4096 |
| Total Bytes Used             |   102400 |
+------------------------------+----------+

SuperVU1: Recompiler cache reset...
Closing plugins...
	Closing DEV9
	Closing FW
	Closing USB
	Closing CDVD
	Closing SPU2
	Closing PAD
	Closing GS
Delete 0 Shaders, 64 Programs, 29 Pipelines
Plugins closed successfully.
Shutting down plugins...
Plugins shutdown successfully.
`

**Comment by prafullpcsx2** [comment]
[crashtagteamtest.zip](https://github.com/PCSX2/pcsx2/files/3008822/crashtagteamtest.zip)
Can you test both these builds and report which one performs better?

**Comment by Kristijan1001** [comment]
With

MicroVU Recompiler 

Test 1
Game Crashes and doesn't even succeed in resetting the cache. Crashes around 80% Fill.
Test2
It works no crashing, no slowing down FPS Wise,No freezing, But it introduces small Stutterings when looking at the coins.

SuperVU

Test 1
Crash
Test 2 
Crash

**Comment by Kristijan1001** [comment]
No Updates on this ?

**Comment by hibye8313** [comment]
Hi, I'm a new comer to this project so apologies if I do or say something stupid.

So I did some testing with this bug and it looks like the problems is the same (I think) as the Scarface problem. The game is constantly making changes to hardcoded constants in the microprogram which causes the microprogram to be recompiled over and over again eventually causing microVU to run out of cache space and reset. The changes are only to the constants the I register is loaded with. I believe Scarface fixes this by testing if the new microprogram has only changes to hardcoded constants and if so will not recompile and will just make the old microprogram access the new constants via memory accesses. If you look at the microprograms that Crash is running the I register settings happens in addresses 0x2bd0 - 0x3ff8 and essentially the rest of the code remains the same, so what I did was check if the new microprogram has exactly the same code in addresses 0x0000 - 0x2cd8 in which case it will consider the 2 microprograms equivalent (besides I register settings). If so it use the same trick as Scarface to get the new I register values directly from memory.

T

his is still pretty hacky because I don't know for sure if there are any other significant changes to the addresses 0x2bd0 - 0x3ff8 that are not I bit settings. In all the microprograms I looked at there was not, or they were other changes in the microprograms that caused it to be recompiled anyway. However, I haven't done enough testing to determine if this will cause other problems in other parts of the game.

One more thing is that this fix seems to work better with doJumpAsSameProgram = true and doJumpCaching = false (both set to the opposite as default, look in microVU_Misc.h for explanation). This prevents the program from being recompiled everytime there is a jump. Since Crash Tag Team seems to jump a lot into addresses in the range 0x2bd0 - 0x3ff8 to load constants this will prevent even more microprograms from being recompiled. When I tested it this did lead to a performance gain and cause the number of microprograms created in a certain part of the game to decrease from about 100 to less than 10. (I put in checks to effectively have doJumpAsSameProgram enabled and doJumpCaching disabled when the hack is enabled).

TLDR: I made a small game fix for Crash Tag Team Racing that seems to resolve the problem with microVU constantly recompiling, running out of cache space, and resetting. Would anyone be able to test this out for me more extensively? If any of the more experienced team members has better suggestion on how to handle this, would be most welcome also =). I also submitted a pull request with the game fix and the gui entry for the fix.

Sidenote: Crash's shadow is not displaying correctly it is showing dark in the whole volume of the shadow as opposed to just on the ground. This was working correctly on a previous version I think. Anyone know why? Maybe need to open an issue for this.

**Comment by prafullpcsx2** [comment]
@hibye8313 I can check it out if you can send me a build. Also if possible have a look at IHRA Drag racing game which seems to have similar problem.

**Comment by MrCK1** [comment]
You don't need to worry about IHRA Drag Racing since I have that one.

@hibye8313 Based on your writeup, are you saying both a modification to the existing Scarface hack AND settings doJumpAsSameProgram to true are required to fix recompilation issues in CTTR?

**Comment by hibye8313** [comment]
@prafullpcsx2 yes just working on pushing the update to Git. Sorry its taking so long I am try to learn how to use git at the same time. It's been a long time since I used it.

@ MrCK1 I dont think so, just the existing Scarface hack is really needed on my computer. In the debug build I noticed a framerate increase from 40fps to about 60 fps between leaving doJumpAsSameProgram = false vs true. There was no significant difference in the release version. I am using my laptop however and so I was using the native resolution. If someone is using upscale resolution etc, it might cause a significant difference. In the devel logs I noticed there were also fewer "equivalent" microprograms created with doJumpAsSameProgram = true (about 10 vs 100 at some parts of the game). But this is expected behavior for that setting and I don't know if it necessarily means better performance. More testing is probably required with other parts of the game. I just decided to enable it since it gave me better performance in the debug version.

I should have that branch pull request pretty soon, sorry about the delay.

**Comment by MrCK1** [comment]
@hibye8313 I went ahead and submitted the PR for you with author credits ;)

I will test later after the Appveyor build is done.

**Comment by hibye8313** [comment]
Great, many thanks MrCK1!

**Comment by Kristijan1001** [comment]
Awesome Job Guys :) Many Thanks 👍 Works Perfectly!

**Comment by Nermal242** [comment]
If i launch the game it only shows black screen

can someone please help me to fix this?

**Comment by tadanokojin** [comment]
Github is not for general support.
Please create an account here https://forums.pcsx2.net/index.php or join our discord.
